Add mp-values

koalive opened the pull request:
Hi! Here is the PR to add the mp-value computation per perturbation (replicates are merged as described in the original paper). I tried to stay as close as possible to the original definition for now but a lot more can be tweaked if needed. Hopefully this integrates correctly with the other operations but I could actually not reuse so many of the `utils` methods as the mp-value is not correlation-based but relies on the Mahalanobis distance. Concretely this means `similarity_melted_df` is not needed for the computation of the mp-value.

Limitations:

- This might still need some linting
- In `cytominer_eval.evaluate` it would be possible to skip the computation of the melted data frame when `operation == "mp_value"`
- I do most of my testing in Jupyter notebooks so I hope I didn't add mistakes when migrating the code to the corresponding files.
- In the methods section of the paper they state "To find S, the covariance matrices are first calculated separately for the treatment and vehicle subsets of P, and then each resulting covariance matrix is weighted by the fraction of total samples in P that it contains. The weighted matrices are summed to produce S." which I did not do for now. This should not affect the result too much anyway.
- The Mahalanobis distance is implemented in `cytominer_eval.operations.util.calculate_mahalanobis`. This could be implemented as separate operation (either in this PR or in another one). I don't think this would make sense to implement directly as a metric as this is not design for pairwise distances.

Cheers.

gwaygenomics commented:
@koalive - I literally just saw this now! Somehow this notification got mixed up somewhere. I will review ASAP. Thank you

gwaygenomics commented:
One quick thing as I review the remainder of the files:

> This might still need some linting

Can you run [`black`](https://github.com/psf/black) and then add the updated files? This automates the linting process and all other scripts have been adjusted

gwaygenomics commented on cytominer_eval/operations/util.py:
```diff
@@ -131,3 +133,87 @@ def get_grit_entry(df: pd.DataFrame, col: str) -> str:
+    # We scale columns by the variance explained
+    if p["rescale_pca"]:
+        pca_array = pca_array * pca.explained_variance_ratio_
+    # NB: this seems useless, as the point of using the Mahalanobis
```
why are we adding it then? 😆

koalive replied:
Why did they do it in the article in the first place? 😅 
It doesn't change the results at all in the current implementation (and a single matrix multiplication doesn't hurt much), so I left the options open for now...

gwaygenomics replied:
cool, sounds good. What dose `NB` mean? We should delete it if it is not adding anything.

koalive replied:
It stands for *nota bene* and is a habit I have to signify that these comments are giving contextual information rather than directly describing the following lines of code. You're right to point out this is not conventional so I'll remove them.

gwaygenomics commented on cytominer_eval/operations/util.py:
```diff
@@ -131,3 +133,87 @@ def get_grit_entry(df: pd.DataFrame, col: str) -> str:
+    p = default_mp_value_parameters()
+    assert (
+        all([x in p.keys() for x in params.keys()])
+    ), "Unknown parameters provided." 
```
```suggestion
    ), "Unknown parameters provided. Only {e} are supported.".format(e=params.keys()) 
```

I didn't test this out, but doing this will make an assertion error more helpful and less frustrating to a user

gwaygenomics commented on cytominer_eval/operations/util.py:
```diff
@@ -131,3 +133,87 @@ def get_grit_entry(df: pd.DataFrame, col: str) -> str:
+    for (k,v) in params.items():
+        p[k] = v
+    
+    merge_df = pert_df.append(control_df)
```
```suggestion
    merge_df = pd.concat([pert_df, control_df], axis="columns").reset_index(drop=True)
```

This is longer, but easier to read what's going on IMO. It also deals with a pandas indexing problem that can get kinda screwy sometimes.

gwaygenomics commented on cytominer_eval/operations/util.py:
```diff
@@ -131,3 +133,87 @@ def get_grit_entry(df: pd.DataFrame, col: str) -> str:
+    control_estimators = DistributionEstimator(pca_array[-control_df.shape[0]:])
+    
+    # Distance between mean of perturbation and control
+    obs = calculate_mahalanobis(pert_df = pca_array[:pert_df.shape[0]],
```
let's use `.iloc` for subsetting numbered indices

koalive replied:
Not sure I got this one. `pca_array` is a `numpy.ndarray` and not a DataFrame. So far I left it as it is.

gwaygenomics commented on cytominer_eval/operations/util.py:
```diff
@@ -131,3 +133,87 @@ def get_grit_entry(df: pd.DataFrame, col: str) -> str:
+    # NB: in the paper's methods section it mentions the covariance used
+    # might be modified to include variation of the perturbation as well.
+    
+    # Permutation test
```
cool! I like the logic here

gwaygenomics commented on cytominer_eval/tests/test_evaluate.py:
```diff
@@ -215,3 +215,75 @@ def test_evaluate_grit():
+                           ]].mp_value >= 0.05)
+    
+    # Strong perturbation should differ from controls
+    assert "PTK2-2" in list(mp_value_results_df[
```
wait, I thought that a low mp_value meant an inactive perturbation. Is this a wrong interpretation?

gwaygenomics commented on cytominer_eval/tests/test_evaluate.py:
```diff
@@ -215,3 +215,75 @@ def test_evaluate_grit():
+        mp_value_results_df[[
+            x in mp_value_gene_control_perts for 
+            x in mp_value_results_df.Metadata_pert_name
+                           ]].mp_value >= 0.05)
```
do you mean `<= 0.05`? It's possible I am interpreting this incorrectly though!

gwaygenomics commented on cytominer_eval/tests/test_evaluate.py:
```diff
@@ -215,3 +215,75 @@ def test_evaluate_grit():
+
+    # Tests on chemical screen dataset
+    mp_value_compound_control_perts = ["DMSO"]
+
```
```suggestion
```
just removing excess empty space

gwaygenomics commented on cytominer_eval/tests/test_evaluate.py:
```diff
@@ -215,3 +215,75 @@ def test_evaluate_grit():
+    )
+          
+    # Negative controls should be "close to themselves"
+    assert all(
```
same confusion about active vs. inactive mp values here

gwaygenomics commented on cytominer_eval/tests/test_operations/test_mp_value.py:
```diff
@@ -0,0 +1,80 @@
+                                control_df = control_df)
+    
+    assert isinstance(maha, float)
+    assert maha >= 0
```
can you assert the actual value? This test can pass in a lot of cases and wouldn't indicate failure well!

gwaygenomics commented on cytominer_eval/tests/test_operations/test_mp_value.py:
```diff
@@ -0,0 +1,80 @@
+    assert result < 1
+    
+    # Distance to itself should be approximately zero
+    # So mp-value should be 1
```
ahhh! I get it. I was interpreting mp-value as mahalanobis distance, instead of as a p value.

I wonder if we should make this more clear in the `evaluate` or `mp_value` docstrings. Perhaps it is already super clear and I just missed it though!

gwaygenomics commented on cytominer_eval/tests/test_transform/test_util.py:
```diff
@@ -116,27 +116,37 @@ def test_check_replicate_groups():
+        elif operation == "mp_value":
             check_replicate_groups(
-                eval_metric=operation, replicate_groups=replicate_group_dict
+                eval_metric=operation, replicate_groups=replicate_groups[0]
```
Why are we using only `Metadata_gene_name` here for replicate groups? It is possible to have multiple columns indicating which perturbations are replicates. `mp_value` doesn't support that?

gwaygenomics replied:
Looking back at `calculate_mp_value()` it should be a quick fix to support a list of columns (it's just used in the groupby).

I looked back at [`grit()`](https://github.com/cytomining/cytominer-eval/blob/56bd9e545d4ce5dea8c2d3897024a4eb241d06db/cytominer_eval/operations/grit.py) and noticed that I  used the same strategy though, so maybe it is ok for now.

If we don't change it, we should add an issue documenting that this could be implemented as an enhancement in both `mp_value` and `grit`.

koalive replied:
I agree, I saw both and had issues making it work with multiple columns so I settled for the single column like for `grit()` for now. This could be added as an enhancement.

gwaygenomics replied:
cool, added in #28

koalive commented:
Hi Greg,
Happy new year to you too!
No worries, thanks for reviewing my PR. With these new commits, I think I addressed your comments and I'll reply to some of them with more details. Overall, regarding the interpretability of these values in comparison to distances, maybe we could simply use the -log10 of the mp-value?
Let me know what you think!

koalive commented on cytominer_eval/operations/util.py:
```diff
@@ -131,3 +133,87 @@ def get_grit_entry(df: pd.DataFrame, col: str) -> str:
     return str(list(entries)[0])
+
+
+class DistributionEstimator:
```
This class would be suitable for other distribution-based distances but for now I agree `MahalanobisEstimator` is more explicit (and this can always be reverted later if needed).

koalive commented on cytominer_eval/operations/util.py:
```diff
@@ -131,3 +133,87 @@ def get_grit_entry(df: pd.DataFrame, col: str) -> str:
+    
+    # Distance between mean of perturbation and control
+    maha = control_estimators.mahalanobis(
+                np.array(np.mean(pert_df, 0)).reshape(1, -1))[0]
```
`DistributionEstimator().mahalanobis()` (which is basically a wrapper of `sklearn.covariance.EmpiricalCovariance.mahalanobis`) can compute distances for multiple points within  a single call. Here we subset it because we compute the distance for a single point and need to access this single value. I think it makes more sense to do it here rather than removing a feature of `DistributionEstimator().mahalanobis()`.

gwaygenomics commented on cytominer_eval/operations/precision_recall.py:
```diff
@@ -11,7 +11,9 @@
 
 def precision_recall(
-    similarity_melted_df: pd.DataFrame, replicate_groups: List[str], k: int,
+    similarity_melted_df: pd.DataFrame,
```
ah, thanks for including these black formatting updates here as well. I must have missed them in my previous contributions

gwaygenomics commented on cytominer_eval/tests/test_operations/test_mp_value.py:
```diff
@@ -0,0 +1,88 @@
+    assert isinstance(maha, float)
+    # NB: the following value is empirically determined
+    # and not theoretically justified but avoids unwanted
+    # changes in the implementation of the Mahalanobis distance
```
👍

gwaygenomics commented on cytominer_eval/tests/test_operations/test_mp_value.py:
```diff
@@ -0,0 +1,88 @@
+    assert "mp_value" in result.columns
+    assert all(result.mp_value <= 1)
+    assert all(result.mp_value >= 0)
+    assert len(np.unique(df[replicate_id])) == len(result)
```
can you add one more test in `calculate_mp_value()`?

We should enter the following statement:

```python
    assert all(
        [x in p.keys() for x in params.keys()]
    ), "Unknown parameters provided. Only {e} are supported.".format(e=p.keys())
    for (k, v) in params.items():
        p[k] = v
```

I think something like this should work:

```python
with pytest.raises(AssertionError) as ae:
    result = mp_value(
        df=df,
        control_perts=control_perts,
        replicate_id=replicate_id,
        features=features,
    )
    assert "Unknown parameters provided. Only" in str(ae.value)
```

but wait! I just realized we won't ever have access to changing parameters since `mp_value()` doesn't include this option in https://github.com/cytomining/cytominer-eval/pull/26/files#diff-470d83233c24bd0089343f7748832f6036c3b7e8422e1aaaebed657fa7c6a725R46.

Ok, one important question then: Do you think it is important to be able to change these parameters?

If not, then I might prefer to add this as a github issue for a future enhancement. If so, then we should at the very least propagate the change to `mp_value()`, and if super duper important, then we need to propagate the change all the way up to `evaluate()`.

gwaygenomics commented on cytominer_eval/operations/mp_value.py:
```diff
@@ -0,0 +1,49 @@
+    assert replicate_id in df.columns, "replicate_id not found in dataframe columns"
+    
+    # Extract features for control rows
+    control_df = df[df[replicate_id].isin(control_perts)][features]
```
bumping this

gwaygenomics reviewed (commented):
One major discussion point about entering and testing mp_value parameters

koalive commented:
Good catch, I added the parametrization of the mp-value computations at the `mp-value` and `evaluate` stages as well and tested for incorrect parameter names in all cases. If I'm not missing anything, this should be ready for merging! 👍

gwaygenomics commented on cytominer_eval/operations/mp_value.py:
```diff
@@ -0,0 +1,56 @@
+    control_perts: List[str],
+    replicate_id: str,
+    features: List[str],
+    params={},
```
```suggestion
    params: dict = {},
```

to keep type hints consistent

gwaygenomics commented on cytominer_eval/operations/util.py:
```diff
@@ -131,3 +138,102 @@ def get_grit_entry(df: pd.DataFrame, col: str) -> str:
+
+
+def calculate_mp_value(
+    pert_df: pd.DataFrame, control_df: pd.DataFrame, params={}
```
```suggestion
    pert_df: pd.DataFrame, control_df: pd.DataFrame, params: dict = {},
```

type hints. Maybe needs a run of black after accepting this commit

gwaygenomics reviewed (commented):
two minor things, then we're good to merge!

gwaygenomics commented:
I'm going to merge. Thanks again @koalive!! I am eager to collaborate in the future, and will keep you posted about manuscript plans
